### ISSUE TITLE
Commit --squash should prompt for a commit message

### ISSUE THREAD

**Issue by KevOrr** [magit/magit#4405]
When creating a `squash!` commit, magit should prompt for a commit message. From the manpage for `git commit`:

> ```
>        --squash=<commit>
>            Construct a commit message for use with rebase --autosquash. The commit message subject line is taken from the specified commit with a prefix of "squash! ". Can be used with additional commit message options
>            (-m/-c/-C/-F). See git-rebase(1) for details.
> ```

When creating a squash commit, you can add additional message lines which will be appended in the suggested commit message when you `rebase --autosquash` later. Here is an example workflow:

```sh
$ git init
$ echo 'initial' >f
$ git add -A; git commit -m 'initial commit'      
[master (root-commit) 7674652] initial commit
 1 file changed, 1 insertion(+)
 create mode 100644 f
$ echo 'feature' >f
$ git add -A; git commit -m 'feature'             
[master 278200d] feature
 1 file changed, 1 insertion(+), 1 deletion(-)
$ echo 'edited' >f 
$ git add -A; git commit --squash HEAD -m 'edited'
[master 57a6fc6] squash! feature
 1 file changed, 1 insertion(+), 1 deletion(-)
$ git log --graph --pretty=format:'%h%n%n%w(0,4,4)%B' | cat
* 57a6fc6
| 
|     squash! feature
|     
|     edited
| 
* 278200d
| 
|     feature
| 
* 7674652
  
      initial commit
$ git rebase --autosquash --interactive HEAD^^
hint: Waiting for your editor to close the file... Waiting for Emacs...
```
At this point the todo list given is
```
pick 278200d feature
squash 57a6fc6 squash! feature
```
After accepting that list, the suggested commit message that pops up is
```
# This is a combination of 2 commits.
# This is the 1st commit message:

feature

# This is the commit message #2:

squash! feature

edited

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
...
```
After accepting that message...
```
[detached HEAD 566cd8e] feature
 Date: Wed Jun 9 11:27:48 2021 -0400
 1 file changed, 1 insertion(+), 1 deletion(-)
Successfully rebased and updated refs/heads/master.
$ git log --graph --pretty=format:'%h%n%n%w(0,4,4)%B' | cat
* 566cd8e
| 
|     feature
|     
|     squash! feature
|     
|     edited
| 
* 7674652
  
      initial commit
```

I think it is important that the user be able to add messages to squash commits. Admittedly, I am relatively new to squash and fixup commits, but I believe the ability to add messages is the main distinction of squash commits vs fixup commits

**Comment by kyleam** [comment]
> I think it is important that the user be able to add messages to squash commits

See `magit-commit-augment` (`c A` by default).

**Comment by KevOrr** [comment]
Hmm, interesting. What is the purpose of `cs` then? If I understand correctly, multiple `cs` commits will just create a commit message like
```
feature

squash! feature

squash! feature
```

**Comment by tarsius** [comment]
> What is the purpose of `c s` then?

Adding to a commit *without* immediately writing some text that describes those additions, while still recording the fact that commit message has to be adjusted eventually, i.e. once you --autosquash.

**Comment by tarsius** [comment]
> Maybe the docs for `magit-commit-squash` could mention `magit-commit-augment`?

I don't know -- did you check the doc-string before opening this issue?

Currently it begins with

> Create a squash commit, without editing the squash message.

Which kinda implies that there is another variant that does

> Create a squash commit, editing the squash message.

or else, why would we mentioned that detail in the second half of the sentence?

Sure it is not immediately obvious that "augment" is the variant that does the latter, but if we cross reference that, then why stop there and not mention fixup too? And the instant variants.

So IMO no, we shouldn't mention it.

**Comment by KevOrr** [comment]
Yes, I checked the docstring, by pressing `?` in magit. In fact it was that docstring which lead me to believe that there was a missing feature here.

I'm not sure it should be assumed that a user who reads "Create a squash commit, without editing the squash message" will
a) assume that there must be a _different_ command which does edit the squash message
b) guess where to find it

> Or else, why would we mentioned that detail in the second half of the sentence

I think it's at the very least confusing that the command called "squash" does not expose the main feature that squash commits have over fixup commits. At the very least, I do not think that the current docs help to communicate, effectively and concisely, where to find this feature.

Here's how I came to the conclusion that this feature didn't yet exist in magit, if that helps:
1. Use instant fixup and instant squash relatively often. Notice that instant squash prompts for a commit message, whereas instant fixup does not.
2. Learn that that's the distinction between squash and fixup, as confirmed by `man git-commit` and `man git-rebase`
3. One day, use (non-instant) squash commit in magit, and notice that it didn't prompt for a commit message, even though `man git-commit` lists that as the key difference
4. Read the docstring and see `without editing the squash message`

Now, I would have discovered this feature if I had tried every single sub-command under `commit` and `rebase`. I just think this is inefficient, and in my estimation before opening this issue, I thought there would be a low probability that this feature would exist under a different name. This is why I suggest that, in order to save future users' time to find this feature, and to save maintainers' and contributors' time in issue reporting, the docs should point the user in the correct direction.
